### DISCUSSION TITLE
Fault Tolerant Services for GRPC

### DISCUSSION THREAD

**Discussion by oheckmann** [k3s-io/k3s#6527] [Q&A]
Setup: I have a five node RPI Cluster connected with K3s (Specifically a PicoCluster 5H with 5 RPI4 8GB) in high Availability mode with embedded DB, 3 master nodes, and 2 worker nodes. I have been running an Edge Server in Kubernetes Pods. replicated across the cluster to make the system more fault tolerant. It is listening on the port 50051. I have a LoadBalancer service which exposes the port to the cluster.

And a client which is sending information to the server by instantiating a grpc channel with the LoadBalancer's address. The client then sends data to the the server. And this whole system works fine when there are no node or pod failures. However, when a node or server pod goes down, the client loses the connection to the server. In order to reconnect, you have to run the client again, when it then reconnects to one of the servers. Now, I am using K3s to try to make the system so that if there is a node failure, everything keeps working with little or no loss of functionality.

Is this something that I can fix by changing the service with exposes the pods, such that it figures out the connections, and if there is a pod failure, it can reroute the channel?

**Comment by brandond** [comment]
> And a client which is sending information to the server by instantiating a grpc channel with the LoadBalancer's address. The client then sends data to the the server. And this whole system works fine when there are no node or pod failures. However, when a node or server pod goes down, the client loses the connection to the server. In order to reconnect, you have to run the client again, when it then reconnects to one of the servers.

It sounds like your client needs to be better about service discovery and retries. If you're running multiple pods, there are ways to discover the service's back-end endpoints via DNS or Kubernetes queries. Similarly, the client should be smart enough to reconnect to the service's IP if it is disconnected.

This all needs to be done in your client though; it's not a Kubernetes issue.
